File: apps/ios/Loonext/Features/Tasks/TaskFormat.swift
import Foundation

/// Pure task display/encoding helpers — the Swift siblings of the web's
/// task-format.ts / task-activity.ts and the Android TaskFormat.kt, kept
/// dependency-free for unit tests. Due dates go amber ONLY when overdue
/// (never a red scare), and only for a not-done task.
///
/// `parseWireTimestamp`, `isOverdue`, and `formatDue` live in
/// Support/Format.swift (shared with the other tabs) — this file adds the
/// tasks-only pieces on top.

let taskTitleMax = 500
let taskDescriptionMax = 5000
let noteBodyMax = 4096
let noteFileMaxBytes = 25 * 1024 * 1024
let noteFilesMax = 10

private let posixLocale = Locale(identifier: "en_US_POSIX")

/// Encode a picked due date as ISO 8601 WITH the zone's UTC offset at that
/// instant (the API requires an offset-bearing string; "Z" only when the zone
/// genuinely is UTC). Example: 2026-07-15 15:00 in America/Toronto →
/// "2026-07-15T15:00:00-04:00". Delegates to the same formatter the due-chip
/// windows use so the two encoders can never drift.
func encodeDueAt(_ date: Date, timeZone: TimeZone = .current) -> String {
    isoOffsetString(date, timeZone: timeZone)
}

/// "today 3:00 PM" / "Jul 8 9:00 AM" for a due-set activity line.
func dueSentenceTime(_ iso: String, now: Date = Date(), calendar: Calendar = .current) -> String {
    guard let date = parseWireTimestamp(iso) else { return "" }
    let formatter = DateFormatter()
    formatter.locale = posixLocale
    formatter.calendar = calendar
    formatter.timeZone = calendar.timeZone
    formatter.dateFormat = "h:mm a"
    let time = formatter.string(from: date)
    if calendar.isDate(date, inSameDayAs: now) { return "today \(time)" }
    formatter.dateFormat = "MMM d"
    return "\(formatter.string(from: date)) \(time)"
}

/// A quiet human sentence for one task_* activity event, ported from the
/// web's taskEventSentence so the clients read identically. `by` is the
/// resolved actor name (fall back to "Loonext" for system); `memberName`
/// resolves the assigned-to user id. Unknown types return nil — skip the row.
func taskEventSentence(
    _ item: TaskActivityItem,
    by: String,
    memberName: (String?) -> String?,
    now: Date = Date(),
    calendar: Calendar = .current
) -> String? {
    func payloadString(_ key: String) -> String? {
        item.payload?[key]?.stringValue
    }
    switch item.type {
    case "task_created":
        return "\(by) turned this into a task"
    case "task_assigned":
        guard let to = payloadString("to_user_id") else {
            return "\(by) unassigned this task"
        }
        if let name = memberName(to) {
            return "\(by) assigned this to \(name)"
        }
        return "\(by) reassigned this task"
    case "task_due_set":
        guard let due = payloadString("due_at") else {
            return "\(by) cleared the due date"
        }
        return "\(by) set the due date to \(dueSentenceTime(due, now: now, calendar: calendar))"
    case "task_deleted":
        return "\(by) removed this task"
    case "task_attachment_added":
        return "\(by) attached a file"
    case "task_attachment_removed":
        return "\(by) removed a file"
    default:
        return nil
    }
}

/// "512 B" / "12 KB" / "3.5 MB" for a file cell; nil size renders nothing.
func formatBytes(_ bytes: Int?) -> String {
    guard let bytes else { return "" }
    if bytes >= 1024 * 1024 {
        return String(format: "%.1f MB", Double(bytes) / 1024.0 / 1024.0)
    }
    if bytes >= 1024 { return "\(bytes / 1024) KB" }
    return "\(bytes) B"
}

File: apps/android/app/src/main/kotlin/com/loonext/android/features/tasks/TaskFormat.kt
package com.loonext.android.features.tasks

import com.loonext.android.core.i18n.AppStrings
import com.loonext.android.core.model.Task
import com.loonext.android.core.model.TaskActivityItem
import java.time.Clock
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter

/**
 * Pure task display/encoding helpers — the Kotlin siblings of the web's
 * task-format.ts and task-activity.ts, kept dependency-free for unit tests.
 * Due dates go amber ONLY when overdue (never a red scare), and only for a
 * not-done task.
 */

const val TASK_TITLE_MAX = 500
const val TASK_DESCRIPTION_MAX = 5000
const val NOTE_BODY_MAX = 4096
const val TASK_SEARCH_MAX = 200

/**
 * Parse a wire timestamp that may be UTC ("…Z"), offset-bearing
 * ("…-04:00" — due_at echoes what clients wrote), or offset-less. Null when
 * unparseable — callers render nothing rather than crash on a new shape.
 */
fun parseInstant(iso: String?): Instant? {
    if (iso == null) return null
    return runCatching { OffsetDateTime.parse(iso).toInstant() }
        .recoverCatching { Instant.parse(iso) }
        .recoverCatching {
            LocalDateTime.parse(iso).atZone(ZoneId.systemDefault()).toInstant()
        }
        .getOrNull()
}

/** A not-done task whose due date is in the past. */
fun isOverdue(task: Task, clock: Clock = Clock.systemDefaultZone()): Boolean {
    if (task.done || task.due_at == null) return false
    val due = parseInstant(task.due_at) ?: return false
    return due.isBefore(clock.instant())
}

/**
 * A short human due label for a chip/cell: "Today", "Tomorrow", "Jul 8",
 * "Jul 8 2027". Null/unparseable due → "" (the caller renders nothing).
 */
fun formatDue(
    dueAt: String?,
    clock: Clock = Clock.systemDefaultZone(),
    // #228: last, and defaulted, so every existing positional call site (and
    // the vectors that pin the English) is untouched while the screens that
    // know the reader's language can pass it.
    locale: String? = null,
): String {
    val instant = parseInstant(dueAt) ?: return ""
    val date = instant.atZone(clock.zone).toLocalDate()
    val today = LocalDate.now(clock)
    return when {
        date == today -> AppStrings.translate(locale, "contactsTasks.today")
        date == today.plusDays(1) -> AppStrings.translate(locale, "contactsTasks.tomorrow")
        date.year == today.year -> date.format(DateTimeFormatter.ofPattern("MMM d"))
        else -> date.format(DateTimeFormatter.ofPattern("MMM d yyyy"))
    }
}

/** "today 3:00 PM" / "Jul 8 9:00 AM" for a due-set activity line. */
fun dueSentenceTime(
    iso: String,
    clock: Clock = Clock.systemDefaultZone(),
    locale: String? = null,
): String {
    val instant = parseInstant(iso) ?: return ""
    val zoned = instant.atZone(clock.zone)
    val time = zoned.format(DateTimeFormatter.ofPattern("h:mm a"))
    return if (zoned.toLocalDate() == LocalDate.now(clock)) {
        AppStrings.translate(locale, "contactsTasks.todayAtTime", mapOf("time" to time))
    } else {
        "${zoned.format(DateTimeFormatter.ofPattern("MMM d"))} $time"
    }
}

/**
 * Encode a picked due date+time as ISO 8601 WITH the zone's UTC offset at
 * that instant (the API requires an offset-bearing string; "Z" only when the
 * zone genuinely is UTC). Example: 2026-07-15 15:00 in America/Toronto →
 * "2026-07-15T15:00:00-04:00".
 */
fun encodeDueAt(local: LocalDateTime, zone: ZoneId): String =
    local.atZone(zone).toOffsetDateTime().format(DUE_AT_FORMAT)

/** Always emits seconds ("…T15:00:00-04:00"); XXX prints Z at offset zero. */
private val DUE_AT_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ssXXX")

/**
 * A quiet human sentence for one task_* activity event, ported from the
 * web's taskEventSentence so the two clients read identically. [by] is the
 * resolved actor name (fall back to "Loonext" for system); [memberName]
 * resolves the assigned-to user id. Unknown types return null — skip the row.
 */
fun taskEventSentence(
    item: TaskActivityItem,
    by: String,
    memberName: (String?) -> String?,
    clock: Clock = Clock.systemDefaultZone(),
    locale: String? = null,
): String? {
    val payload = item.payload
    fun payloadString(key: String): String? =
        (payload?.get(key) as? kotlinx.serialization.json.JsonPrimitive)
            ?.takeIf { it.isString }?.content

    fun say(key: String, vararg vars: Pair<String, String>) =
        AppStrings.translate(locale, key, mapOf("by" to by) + vars.toMap())

    return when (item.type) {
        "task_created" -> say("contactsTasks.activityCreated")
        "task_assigned" -> {
            val to = payloadString("to_user_id")
            when {
                to == null -> say("contactsTasks.activityUnassigned")
                else -> memberName(to)
                    ?.let { say("contactsTasks.activityAssignedTo", "who" to it) }
                    ?: say("contactsTasks.activityReassigned")
            }
        }

        "task_due_set" -> {
            val due = payloadString("due_at")
            if (due == null) {
                say("contactsTasks.activityDueCleared")
            } else {
                say(
                    "contactsTasks.activityDueSet",
                    "when" to dueSentenceTime(due, clock, locale),
                )
            }
        }

        "task_deleted" -> say("contactsTasks.activityDeleted")
        "task_attachment_added" -> say("contactsTasks.activityAttached")
        "task_attachment_removed" -> say("contactsTasks.activityAttachmentRemoved")
        else -> null
    }
}
